[1/4][DataPipe] Properly cleanup unclosed files within generator function

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #89974 [4/4][DataPipe] Remove iterator depletion in Zipper
* **#89973 [1/4][DataPipe] Properly cleanup unclosed files within generator function**

There is a case that `file.close` is never called because when generator function has never reached to the end. A simple example would be `zip` two datepipes with different length. The longer DataPipe would never reach the end of generator and then it will be cleaned up by `gc`. So, the line of `file.close` is not executed. (This is the reason that Vitaly has to create this [hack](https://github.com/pytorch/pytorch/blob/4451eb24e6287dff62ff8a7ec0eda6a6998807b0/torch/utils/data/datapipes/iter/combining.py#L573-L583) to retrieve all remaining data to make sure generator function is fully executed)

However, this hack introduces another problem where an infinite datapipe would make `zip` never end as it would try to deplete the infinite iterator. See: https://github.com/pytorch/data/issues/865

So, in this PR, I am adding a `try-finally` clause to make sure the `file.close` is always executed during the destruction of `generator` object. Then, we don't need the hack within `zip` any more.

Differential Revision: [D41699470](https://our.internmc.facebook.com/intern/diff/D41699470)

ejguan committed:
[2/4][DataPipe] Properly cleanup unclosed files within generator function

[ghstack-poisoned]

wenleix commented on torch/utils/data/datapipes/iter/combining.py:
curious: why we cannot wait for `self.__del__` being triggered? (looks like the clean up is the same).

Or, does that makes sense to refactor some `_cleanup` and make both `__del__` and here call it? ~

ejguan replied:
> curious: why we cannot wait for `self.__del__` being triggered? (looks like the clean up is the same).

The main problem is that `__del__` is not always executed in Python.


> does that makes sense to refactor some `_cleanup` and make both `__del__` and here call it? ~

Yeah, it makes sense to me.

NivekT replied:
Agree that we can refactor to `_cleanup` but I'm also okay with this as it is since it is only called twice

wenleix commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -345,9 +347,9 @@ def autoclose(self):
+        self.close_on_last_child = True
         if self.child_counter == 0:
             self.close()
-        self.close_on_last_child = True
```
hmm.. why originally set `close_on_last_child` after closing it...

wenleix replied:
to clarify, i mean the new implementation looks correct, just don't understand the original implementation~

wenleix reviewed (commented):
LGTM. Would like @NivekT to also take a look~

NivekT reviewed (approved):
LGTM! It'd be good if we have a unit test to check that things are closed, but not the most essential.

ejguan committed:
Update on "[1/4][DataPipe] Properly cleanup unclosed files within generator function"


There is a case that `file.close` is never called because when generator function has never reached to the end. A simple example would be `zip` two datepipes with different length. The longer DataPipe would never reach the end of generator and then it will be cleaned up by `gc`. So, the line of `file.close` is not executed. (This is the reason that Vitaly has to create this [hack](https://github.com/pytorch/pytorch/blob/4451eb24e6287dff62ff8a7ec0eda6a6998807b0/torch/utils/data/datapipes/iter/combining.py#L573-L583) to retrieve all remaining data to make sure generator function is fully executed)

However, this hack introduces another problem where an infinite datapipe would make `zip` never end as it would try to deplete the infinite iterator. See: https://github.com/pytorch/data/issues/865

So, in this PR, I am adding a `try-finally` clause to make sure the `file.close` is always executed during the destruction of `generator` object. Then, we don't need the hack within `zip` any more.

[ghstack-poisoned]

ejguan committed:
Update on "[1/4][DataPipe] Properly cleanup unclosed files within generator function"


There is a case that `file.close` is never called because when generator function has never reached to the end. A simple example would be `zip` two datepipes with different length. The longer DataPipe would never reach the end of generator and then it will be cleaned up by `gc`. So, the line of `file.close` is not executed. (This is the reason that Vitaly has to create this [hack](https://github.com/pytorch/pytorch/blob/4451eb24e6287dff62ff8a7ec0eda6a6998807b0/torch/utils/data/datapipes/iter/combining.py#L573-L583) to retrieve all remaining data to make sure generator function is fully executed)

However, this hack introduces another problem where an infinite datapipe would make `zip` never end as it would try to deplete the infinite iterator. See: https://github.com/pytorch/data/issues/865

So, in this PR, I am adding a `try-finally` clause to make sure the `file.close` is always executed during the destruction of `generator` object. Then, we don't need the hack within `zip` any more.

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41699470).

ejguan commented:
This is the proper fix for https://github.com/pytorch/data/issues/436 that we have been trying to resolve.
